Pass by const ref instead of by value in StableIValue from

janeyx99 opened the pull request:
I realize I was passing stable::Tensors by value (thus making a copy every time) which is not what I want from the `from` function that converts Ts to StableIValues. `from` should not mutate the input and should be read-only.

I asked an LLM whether this is API BC breaking (with an intuition that it shouldn't be), and it said no, cuz:
1. "Passing by const reference is more permissive than passing by value. e.g., if T is a type that has a deleted or inaccessible copy constructor (e.g., std::unique_ptr), the original code would have been invalid, while the new code would be valid." Nice. We are good with additive.
2. We didn't modify the original input before (cuz we took a copy) and we don't now (cuz we promise const).

Update: The LLM failed to mention primitives, with which we should not pass references around, so we are only changing the signatures of std::optional<T> and stable::Tensor

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #156126
* #155977
* #155367

janeyx99 committed:
Pass by const ref instead of by value in StableIValue from

[ghstack-poisoned]

swolchok commented on torch/csrc/stable/library.h:
```diff
@@ -31,7 +31,7 @@ namespace detail {
 template <typename T>
 struct FromImpl {
-  static StableIValue call(T val) {
+  static StableIValue call(const T& val) {
```
> wrong

by wrong, I mean "needlessly inefficient", though if the function is inlined then it doesn't matter

janeyx99 replied:
thanks for real human 100% understanding and knowledge--done!

janeyx99 committed:
Update on "Pass by const ref instead of by value in StableIValue from"

I realize I was passing stable::Tensors by value (thus making a copy every time) which is not what I want from the `from` function that converts Ts to StableIValues. `from` should not mutate the input and should be read-only.

I asked an LLM whether this is API BC breaking (with an intuition that it shouldn't be), and it said no, cuz:
1. "Passing by const reference is more permissive than passing by value. e.g., if T is a type that has a deleted or inaccessible copy constructor (e.g., std::unique_ptr), the original code would have been invalid, while the new code would be valid." Nice. We are good with additive.
2. We didn't modify the original input before (cuz we took a copy) and we don't now (cuz we promise const).




[ghstack-poisoned]

swolchok reviewed (commented):
btw, it's not strictly relevant to this PR but from() itself takes a T, which isn't great. I think you can fix that by overloading from() on const torch::stable::Tensor&, just be careful if there are any possible implicit conversions. https://godbolt.org/z/aT94nhann

swolchok commented on torch/csrc/stable/library.h:
```diff
@@ -88,7 +88,7 @@ struct FromImpl<std::nullopt_t> {
 template <typename T>
 struct FromImpl<std::optional<T>> {
-  static StableIValue call(std::optional<T> val) {
+  static StableIValue call(const std::optional<T>& val) {
```
if T can be something other than a cheaply copyable primitive then fine, otherwise you should revert this one too.

janeyx99 committed:
Update on "Pass by const ref instead of by value in StableIValue from"

I realize I was passing stable::Tensors by value (thus making a copy every time) which is not what I want from the `from` function that converts Ts to StableIValues. `from` should not mutate the input and should be read-only.

I asked an LLM whether this is API BC breaking (with an intuition that it shouldn't be), and it said no, cuz:
1. "Passing by const reference is more permissive than passing by value. e.g., if T is a type that has a deleted or inaccessible copy constructor (e.g., std::unique_ptr), the original code would have been invalid, while the new code would be valid." Nice. We are good with additive.
2. We didn't modify the original input before (cuz we took a copy) and we don't now (cuz we promise const).




[ghstack-poisoned]

janeyx99 committed:
Update on "Pass by const ref instead of by value in StableIValue from"

I realize I was passing stable::Tensors by value (thus making a copy every time) which is not what I want from the `from` function that converts Ts to StableIValues. `from` should not mutate the input and should be read-only.

I asked an LLM whether this is API BC breaking (with an intuition that it shouldn't be), and it said no, cuz:
1. "Passing by const reference is more permissive than passing by value. e.g., if T is a type that has a deleted or inaccessible copy constructor (e.g., std::unique_ptr), the original code would have been invalid, while the new code would be valid." Nice. We are good with additive.
2. We didn't modify the original input before (cuz we took a copy) and we don't now (cuz we promise const).

Update: The LLM failed to mention primitives, with which we should not pass references around, so we are only changing the signatures of std::optional<T> and stable::Tensor




[ghstack-poisoned]

janeyx99 committed:
Update on "Pass by const ref instead of by value in StableIValue from"

I realize I was passing stable::Tensors by value (thus making a copy every time) which is not what I want from the `from` function that converts Ts to StableIValues. `from` should not mutate the input and should be read-only.

I asked an LLM whether this is API BC breaking (with an intuition that it shouldn't be), and it said no, cuz:
1. "Passing by const reference is more permissive than passing by value. e.g., if T is a type that has a deleted or inaccessible copy constructor (e.g., std::unique_ptr), the original code would have been invalid, while the new code would be valid." Nice. We are good with additive.
2. We didn't modify the original input before (cuz we took a copy) and we don't now (cuz we promise const).

Update: The LLM failed to mention primitives, with which we should not pass references around, so we are only changing the signatures of std::optional<T> and stable::Tensor




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
